src/extra/recipes/cookiebox.py

```diff
@@ -0,0 +1,1047 @@
+from typing import Optional, Union, Dict, List, Tuple, Any
+from dataclasses import dataclass, asdict, is_dataclass
+
+from .base import BaseCalibration
+
+import numpy as np
+from extra_data import open_run, by_id, DataCollection
+import h5py
+import xarray as xr
+import pandas as pd
+
+import itertools
+from concurrent.futures import ThreadPoolExecutor, ProcessPoolExecutor
+
+from extra.components import Scan, AdqRawChannel, XrayPulses, XGM
+
+from lmfit.models import GaussianModel, LinearModel, ConstantModel
+from lmfit.model import ModelResult
+import scipy
+from scipy.stats import linregress
+from scipy.optimize import minimize_scalar
+from functools import partial
+from scipy.signal import fftconvolve
+from scipy.interpolate import CubicSpline
+from scipy.signal import kaiserord, filtfilt, firwin
+
+
+@dataclass
+class TofFitResult:
+    """Class keeping track of fit in a single Tof."""
+    energy: np.ndarray
+    mu: np.ndarray
+    sigma: np.ndarray
+    A: np.ndarray
+    Aa: np.ndarray
+    offset: np.ndarray
+
+
+def search_offset(trace: np.ndarray, sigma: float=20):
+    """
+    Find highest peaks in the 1D trace.
+    """
+    axis = np.arange(len(trace))
+    gaussian = np.exp(-0.5*(axis - np.mean(axis))**2/sigma**2)
+    gaussian /= np.sum(gaussian, axis=0, keepdims=True)
+    # apply it to the data
+    smoothened = fftconvolve(trace, gaussian, mode="same", axes=0)
+    peak_idx = np.argmax(smoothened)
+    return peak_idx - 200
+
+def search_roi(roi: np.ndarray):
+    """
+    Find highest peaks in the 1D trace.
+    """
+    p, _ = scipy.signal.find_peaks(roi, prominence=(0.25*np.amax(roi), None))
+    return p
+
+def model(ts,c,e0,t0):
+    """
+    Model to fit energy as a function of the sample axis.
+    """
+    # e = e0 + c/(ts - t0)**2
+    # ts = np.sqrt(c/(e - e0)) + t0
+    # dt/de = d/de [ sqrt(c/e) ] = d/du sqrt(u) d/de [c/e] = -0.5*c/sqrt(c/e) * 1/(e-e0)^2 
+    # dt/de = -0.5*c/(np.sqrt(c/(e - e0)))/(e - e0)**2
+    # before transmission, after time -> energy interpolation:
+    # multiply peak sizes by dt/de
+    return e0+c/(ts-t0)**2
+
+# e = e0 + c/(ts-t0)^2
+# 1/(e - e0)
+def lin_fit(ts,energies,t0):
+    """
+    Fit model using linear regression for fixed t0.
+    """
+    res = linregress(1/np.asarray(ts-t0)**2,np.asarray(energies))
+    c = res.slope
+    e0 = res.intercept
+    return c,e0,t0
+
+def norm_diff_t0(ts,energies,t0):
+    #print(ts,'fi',energies)
```

Comment:
Nitpicking, I think this can be removed (same for the other print statements).

Reply:
Agreed!